lib/topic_upload_security_manager.rb

```diff
@@ -89,17 +89,17 @@ def run
   private
 
   def posts_owning_uploads
-    Post.where(topic_id: @topic.id).joins("INNER JOIN uploads ON access_control_post_id = posts.id")
+    Post.where(topic_id: @topic.id, id: Upload.select(:access_control_post_id))
```

Comment:
@SamSaffron @danielwaterworth @pmusaraj  hold on...we just merged this but this line and down below where we do `Upload.where(access_control_post_id: nil)` are quite confusing -- this and `Upload.select(:access_control_post_id)` could be returning thousands and thousands of IDs...

This should be something more like:

```ruby
Post.where(
  topic_id: @topic.id,
  id: Upload.where("access_control_post_id IS NOT NULL AND access_control_post_id IN (SELECT id FROM posts WHERE topic_id = #{@topic.id})").select(:access_control_post_id)
)
```

And then the below one like:

```ruby
UploadReference
  .joins(:upload)
  .where(target_type: "Post")
  .where("target_id IN (SELECT id FROM posts WHERE topic_id = #{topic.id})")
  .where("uploads.access_control_post_id IS NULL")
  .select(:target_id)
```

Otherwise it feels like this can blow up very fast since we are not limiting the scope of the upload queries. Am I missing something?

Reply:
`select` doesn't run the query. This does get turned into a subquery, i.e. `id IN (SELECT access_control_post_id FROM uploads)`.

Reply:
Yes I saw, but isn't that also wasteful to be doing such a massive `id IN (X)` where `X` can be thousands of records? Wouldn't it be better to limit the scope of the `Upload` and `UploadReference` queries with the information we already know?

Reply:
Not necessarily, but it's difficult to answer questions like this without a query plan so:

```
                                                     QUERY PLAN                 
                                    
--------------------------------------------------------------------------------------------------------------------
 Nested Loop Semi Join  (cost=0.85..29.50 rows=1 width=966)
   ->  Index Scan using index_posts_on_topic_id_and_created_at on posts  (cost=0
.42..18.68 rows=17 width=966)
         Index Cond: (topic_id = 5)
   ->  Index Only Scan using index_uploads_on_access_control_post_id on uploads 
 (cost=0.42..12.36 rows=63 width=8)
         Index Cond: (access_control_post_id = posts.id)
(5 rows)
```

`Nested Loop Semi Join` iterates over the first set and probes into the second. So, it's doing a good job. It's filtering out posts by doing a lookup per post in the `uploads (access_control_post_id)` index. With this plan, adding extra conditions on the uploads subquery would actually slow it down, because it would have to look in more places.

Reply:
Okay I am satisfied! I just thought I would call it out because I wasn't sure and it looked weird to me. Thanks for doing the query plan :)